R/test-compiled-code.R

```diff
@@ -21,10 +26,66 @@ expect_cpp_tests_pass <- function(package) {
     }
   )
 
-  # Drop first line of output (it's jut a '####' delimiter)
-  info <- paste(output[-1], collapse = "\n")
+  report <- xml2::read_xml(paste(output, collapse = "\n"))
+  xml2::write_xml(report, "test.xml")
+
+  contexts <- xml2::xml_find_all(report, "//TestCase")
+
+  for (context in contexts) {
+    context_name <- sub(" [|][^|]+$", "", xml2::xml_attr(context, "name"))
+
+    get_reporter()$start_context(context = context_name)
+
+    tests <- xml2::xml_find_all(context, "./Section")
+    for (test in tests) {
+      test_name <- xml2::xml_attr(test, "name")
+
+      result <- xml2::xml_find_first(test, "./OverallResults")
+      successes <- as.integer(xml2::xml_attr(result, "successes"))
+
+      get_reporter()$start_test(context = context_name, test = test_name)
+
+      for (i in seq_len(successes)) {
+        exp <- expectation("success", "")
+        exp$test <- test_name
+        get_reporter()$add_result(context = context_name, test = test_name, result = exp)
+      }
+
+      failures <- xml2::xml_find_all(test, "./Expression")
+      for (failure in failures) {
+        org <- xml2::xml_find_first(failure, "Original")
+        org_text <- xml2::xml_text(org, trim = TRUE)
+
+        filename <- xml2::xml_attr(failure, "filename")
 
-  expect(tests_passed, paste("C++ unit tests:", info, sep = "\n"))
+        line <- xml2::xml_attr(failure, "line")
+        failure_srcref <- srcref(srcfile(file.path("src", filename)), c(line, line, 1, 1))
+
+        exp <- expectation("failure", org_text, srcref = failure_srcref)
+        exp$test <- test_name
```

Comment:
This is slightly awkward, but the reporters expect a test file as a list element. It is normally added by `test_code()` at https://github.com/r-lib/testthat/blob/c5daf5ba21692a38de2c3068c5b8b8bb741ffab6/R/test-that.R#L59